### ISSUE TITLE
st-pipe function is being called twice in a row

### ISSUE THREAD

**Issue by yanivkalfa** [lorenzofox3/Smart-Table#329]
Not really sure whats going on but st-pipe function is being called twice in a row which isnt ideal,

**Comment by lorenzofox3** [comment]
Most of the directives trigger a call to pipe function whenever the table state change, normally it is one shot but there could be side effects. Can you show me what your table look like ?

**Comment by yanivkalfa** [comment]
Please take a look at: 
http://plnkr.co/edit/EDdItB54ZXKSVZkOkSF7?p=preview

I've built this base on you ajax plugin example.

**Comment by lorenzofox3** [comment]
yep you are right, I think the pagination directive when linking changes somehow the tableState define [here](https://github.com/lorenzofox3/Smart-Table/blob/master/src/stTable.js#L10-L16) which triggers a call to pipe function (the first one is from the post link function of the main directive). I need to investigate a bit more. Meanwhile, as a workaround, I suggest you cache you http request or check manually in you custom pipe function to avoid duplicate call for what should be the same tableStae

**Comment by yanivkalfa** [comment]
meanwhile as a solution we have a flag in place.

**Comment by smebberson** [comment]
I'm experiencing this too!

**Comment by alainlecluse** [comment]
Samething here.

**Comment by alainlecluse** [comment]
Just found out that it's not just st-pagination but also st-sort-default. @ version 2.0.1.
My tables with pipe calls the function 6 times.

**Comment by smajl** [comment]
Is 2.0.1 even working with custom pipe calls? Can't get it to work at all.. Pipe function gets called too many times, data duplicating, digests limit reached, horrible.

**Comment by smajl** [comment]
Confirmed. pagination and st-sort-default do this. :(

**Comment by yanivkalfa** [comment]
@lorenzofox3,  what will one do if one desires to override your default pipe / sort by behavior ?  how can i extend on your controller ? 

Say i want to add a multi-sort client side, i need a way to override pipe. overriding sortBy is easy because you have only tableState variable to worry about and you have a method to get that. but pipe has several others variables...

**Comment by Darksoulsong** [comment]
Issue is still happening, same use case.

**Comment by rockccf** [comment]
Yes, using version 2.1.6 and it's still happening.
When is this going to be fixed?
Thanks.

**Comment by rafaelhz** [comment]
It still happening when included st-pagination. (version 2.1.7)

**Comment by tistvan** [comment]
I just updated smartTable to 2.1.8 and i still have 2 calls. The first without sort and the second with sort. I am using st-sort-default directive.

**Comment by Marcrazyness** [comment]
Removing `st-sort-default="true"` on the table column fixed my issue.

**Comment by JinPink** [comment]
I had debugged this issue and resolved it.
**Reason:**
the code at the end of 'stPagination' directive(v2.1.8)
```
490        if (!ctrl.tableState().pagination.number) {
491          ctrl.slice(0, scope.stItemsByPage);
492        }
```
**Solution:**
initialise tableState.pagination.number in your pipe service.
`
tableState.pagination.number = vm.numPerPage;
`

**Comment by csabaujvari** [comment]
I updated smart table to the latest version, but still having this issue with st-sort-default directive. 

I found an alternative solution. A bit hackish, sorry :( 

>         vm.callServer = function (tableState) {
            if(!vm.firstLoadIsDone) {
                vm.firstLoadIsDone = true;
                return;
            }
            ....
        }

**Comment by newbro** [comment]
I was trying to implement persistence tableState for ajax call and seems restoring tableState will cause pipe function being called twice as mention here. Is there any fix in plan?

**Comment by swatifursule** [comment]
What is workaround or fix for this?

**Comment by rockccf** [comment]
Refer to MrWook's reply in issue 734,

Probably the problem occurs when you use sort with the pipe directive and like i said in #388

I used my previous answer for a while and find out that it destroy my Table page caching system so I decided to increase the pipe delay to 400ms. I really don't like this because i see this delay maybe not the customer but i see it.
Now i was thinking about the reason behind it why the increase of the delay fixed this issue.
I decided to look throw the SmartTable.js file and i saw that the sorting have a delay to so i decrease the delay from the sorting and set the delay from the pipe right above it.
The pipe delay need to be higher than the sort delay
To provide you with some more help:
app.config([ 'stConfig', function(stConfig) { stConfig.sort.delay = 100; stConfig.pipe.delay = 200; }]);
